Midichlorians/Server/server.h

```diff
@@ -0,0 +1,229 @@
+
+#ifndef CURRENT_MIDICHLORIANS_SERVER_H
+#define CURRENT_MIDICHLORIANS_SERVER_H
+
+#include "schema.h"
+
+#include "../../TypeSystem/Serialization/json.h"
+
+#include "../../Blocks/HTTP/api.h"
+#include "../../Bricks/strings/strings.h"
+#include "../../Bricks/time/chrono.h"
+
+namespace current {
+namespace midichlorians {
+namespace server {
+
+template <class LOG_ENTRY_CONSUMER>
+class MidichloriansHTTPServer {
+ public:
+  MidichloriansHTTPServer(int http_port,
+                          LOG_ENTRY_CONSUMER& log_entry_consumer,
+                          std::chrono::microseconds tick_interval_us,
+                          const std::string& route = "/log",
+                          const std::string& response_text = "OK\n")
+      : http_port_(http_port),
+        log_entry_consumer_(log_entry_consumer),
+        route_(route),
+        response_text_(response_text),
+        tick_interval_us_(tick_interval_us),
+        send_ticks_(tick_interval_us_.count() > 0),
+        last_event_t_(0u),
+        events_pushed_(0u),
+        timer_thread_(&MidichloriansHTTPServer::TimerThreadFunction, this),
+        routes_(HTTP(http_port_).Register(route_, [this](Request r) { ProcessRequest(std::move(r)); })) {}
+
+  ~MidichloriansHTTPServer() {
+    send_ticks_ = false;
+    timer_thread_.join();
+  }
+
+  void Join() { HTTP(http_port_).Join(); }
+
+  size_t EventsPushed() const { return events_pushed_; }
+
+ private:
+  void ProcessRequest(Request r) {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (r.method == "POST" && !r.body.empty()) {
+        // POST request could be either iOS or web event.
+        // In case of iOS there could be a bunch of events in the body.
+        std::istringstream iss(r.body);
+        std::string line;
+        while (std::getline(iss, line)) {
+          if (!BodyLineAsiOSEvent(line)) {
+            if (!RequestAsWebEvent(r)) {
+              LogInvalidRequest(r);
+            }
+          }
+        }
+      } else if (r.method == "GET" || r.method == "HEAD") {
+        // GET and HEAD requests could be only web events.
+        if (!RequestAsWebEvent(r)) {
+          LogInvalidRequest(r);
+        }
+      } else {
+        LogInvalidRequest(r);
+      }
+    }
+    r(response_text_);
+  }
+
+  bool BodyLineAsiOSEvent(const std::string& line) {
+    using namespace current::midichlorians::ios;
+
+    Variant<T_IOS_EVENTS> ios_event;
+    try {
+      ios_event = ParseJSON<Variant<T_IOS_EVENTS>>(line);
+    } catch (const InvalidJSONException&) {
+      return false;
+    }
+
+    PassLogEntryToConsumer(EventLogEntry(current::time::Now(), std::move(ios_event)));
+    return true;
+  }
+
+  bool RequestAsWebEvent(const Request& r) {
+    using namespace current::midichlorians::web;
+
+    std::map<std::string, std::string> post_q;  // Query parameters extracted from POST body.
+    const std::map<std::string, std::string>* q;
```

Comment:
A neat trick is `const std::map<std::string, std::string>& q = [...](...){}();`.